catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/DumpCommand.java

```diff
@@ -407,13 +413,139 @@ protected void printStatus(long count) {
+      throw new CatalogTransformerException(
+          String.format(
+              "Error occurred when initializing/closing ZipOutputStream with path %s.", filePath),
+          e);
+    }
+  }
+
+  private void writeMetacardToZip(ZipOutputStream zipOutputStream, Metacard metacard) {
+    InputStream inputStream = null;
+
+    try (FileBackedOutputStream fileBackedOutputStream = new FileBackedOutputStream(BUFFER_SIZE);
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileBackedOutputStream)) {
+
+      ZipEntry zipEntry = new ZipEntry(METACARD_PATH + metacard.getId());
+      zipOutputStream.putNextEntry(zipEntry);
+
+      objectOutputStream.writeObject(new MetacardImpl(metacard));
+      inputStream = fileBackedOutputStream.asByteSource().openStream();
+
+      IOUtils.copy(inputStream, zipOutputStream);
+    } catch (IOException e) {
+      LOGGER.debug("Failed to add metacard with id {}.", metacard.getId(), e);
+    } finally {
+      try {
+        if (inputStream != null) {
+          inputStream.close();
+        }
+      } catch (IOException e) {
+        LOGGER.debug("Failed to close input stream", e);
+      }
+    }
+  }
+
+  private boolean hasLocalResources(Metacard metacard) {
+    URI uri = metacard.getResourceURI();
+    return (uri != null && ContentItem.CONTENT_SCHEME.equals(uri.getScheme()));
+  }
+
+  private Map<String, Resource> getAllMetacardContent(Metacard metacard) {
+    Map<String, Resource> resourceMap = new HashMap<>();
+    Attribute attribute = metacard.getAttribute(Metacard.DERIVED_RESOURCE_URI);
+
+    if (attribute != null) {
+      List<Serializable> serializables = attribute.getValues();
+      serializables.forEach(
+          serializable -> {
+            String fragment = CONTENT + File.separator;
+            URI uri = null;
+            try {
+              uri = new URI((String) serializable);
+              String derivedResourceFragment = uri.getFragment();
+              if (ContentItem.CONTENT_SCHEME.equals(uri.getScheme())
+                  && StringUtils.isNotBlank(derivedResourceFragment)) {
+                fragment += derivedResourceFragment + File.separator;
+                Resource resource = getResource(metacard);
+                if (resource != null) {
+                  resourceMap.put(
+                      fragment + uri.getSchemeSpecificPart() + "-" + resource.getName(), resource);
+                }
+              }
+            } catch (URISyntaxException e) {
+              LOGGER.debug(
+                  "Invalid Derived Resource URI Syntax for metacard : {}", metacard.getId(), e);
+            }
+          });
+    }
+
+    URI resourceUri = metacard.getResourceURI();
+
+    Resource resource = getResource(metacard);
+
+    if (resource != null) {
+      resourceMap.put(
+          CONTENT + File.separator + resourceUri.getSchemeSpecificPart() + "-" + resource.getName(),
+          resource);
+    }
+
+    return resourceMap;
+  }
```

Comment:
okay now like i mentioned- i started to see if i could tweak one line, then kept on changing and tweaking and just basically ended up writing a flatter version of this which may or may not provide any value -- so i'll post it here just for posterity and enjoyment 🤷‍♂️ Don't feel like you need to change anything unless you feel really called to do so 😛 

```java
  private Map<String, Resource> getAllMetacardContent(Metacard metacard) {
    Attribute attribute = metacard.getAttribute(Metacard.DERIVED_RESOURCE_URI);
    Map<String, Resource> resourceMap = getResourceMap(metacard, attribute);

    URI resourceUri = metacard.getResourceURI();

    Resource resource = getResource(metacard);

    if (resource != null) {
      resourceMap.put(
          CONTENT + File.separator + resourceUri.getSchemeSpecificPart() + "-" + resource.getName(),
          resource);
    }

    return resourceMap;
  }

  private Map<String, Resource> getResourceMap(Metacard metacard, Attribute attribute) {
    if (attribute == null) {
      return new HashMap<>();
    }
    return attribute
        .getValues()
        .stream()
        .map(serializable -> getDerivedResources(metacard, serializable))
        .filter(Objects::nonNull)
        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (first, next) -> first));
  }

  @Nullable
  private Map.Entry<String, Resource> getDerivedResources(
      Metacard metacard, Serializable serializable) {
    if (!(serializable instanceof String)) {
      LOGGER.debug(
          "Input ({}) should have been a string but was a {}",
          serializable,
          serializable.getClass());
      return null;
    }

    URI uri = null;
    try {
      uri = new URI((String) serializable);
    } catch (URISyntaxException e) {
      LOGGER.debug("Invalid Derived Resource URI Syntax for metacard : {}", metacard.getId(), e);
      return null;
    }

    String derivedResourceFragment = uri.getFragment();
    if (!ContentItem.CONTENT_SCHEME.equals(uri.getScheme())
        || StringUtils.isBlank(derivedResourceFragment)) {
      return null;
    }

    String fragment = CONTENT + File.separator + derivedResourceFragment + File.separator;
    Resource resource = getResource(metacard);
    if (resource == null) {
      return null;
    }

    return new SimpleEntry<>(
        fragment + uri.getSchemeSpecificPart() + "-" + resource.getName(), resource);
  }

```

Reply:
Nice this is definitely more readable. I tested it out and it worked perfectly too!